instrumentation/kafka/src/test/java/brave/kafka/ITTracingKafka.java

```diff
@@ -0,0 +1,161 @@
+package brave.kafka;
+
+import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
+
+import brave.Tracing;
+import brave.sampler.Sampler;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.Future;
+import kafka.admin.AdminUtils;
+import kafka.server.KafkaConfig;
+import kafka.server.KafkaServer;
+import kafka.utils.TestUtils;
+import kafka.utils.ZKStringSerializer$;
+import kafka.utils.ZkUtils;
+import kafka.zk.EmbeddedZookeeper;
+import org.I0Itec.zkclient.ZkClient;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.utils.Time;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import zipkin.Span;
+
+public class ITTracingKafka {
+
+    private static final String LOCALHOST = "127.0.0.1";
+    private static final String KAFKA_PORT = "9092";
+
+    private static final String TEST_TOPIC = "myTopic";
+    private static final String TEST_KEY = "foo";
+    private static final String TEST_VALUE = "bar";
+
+    private KafkaServer kafkaServer;
+    private ZkUtils zkUtils;
+
+    private Tracing consumerTracing;
+    private Tracing producerTracing;
+
+    private ConcurrentLinkedDeque<Span> consumerSpans = new ConcurrentLinkedDeque<>();
+    private ConcurrentLinkedDeque<Span> producerSpans = new ConcurrentLinkedDeque<>();
+
+    @Before
+    public void setUp() throws Exception {
+        consumerTracing = Tracing.newBuilder()
+            .reporter(consumerSpans::add)
+            .sampler(Sampler.ALWAYS_SAMPLE)
+            .build();
+        producerTracing = Tracing.newBuilder()
+            .reporter(producerSpans::add)
+            .sampler(Sampler.ALWAYS_SAMPLE)
+            .build();
+
+        initKafkaServer();
+    }
+
+    @After
+    public void closeKafkaServer() {
+        kafkaServer.shutdown();
+    }
+
+    @Test
+    public void produce_and_consume_kafka_message() throws Exception {
+        TracingProducer<String, String> tracingProducer = createTracingProducer();
+        TracingConsumer<String, String> tracingConsumer = createTracingConsumer();
+
+        Future<RecordMetadata> send = tracingProducer.send(new ProducerRecord<>(TEST_TOPIC, TEST_KEY, TEST_VALUE));
+        // Block for synchronous send
+        send.get();
+
+        ConsumerRecords<String, String> records = tracingConsumer.poll(1000);
+        // Oddly we need to poll twice to retrieve content
```

Comment:
I think that poll needs to be called in a loop that exits once a record is returned or some safeguard timeout is reached. Calling poll is not guaranteed to return records when called, even if there are messages on the broker that could be consumed. See http://grokbase.com/p/kafka/users/155mqpwf3n/consumer-poll-returns-no-records-unless-called-more-than-once-why for more details. That post is a couple years old, but I'm pretty sure the advice is still accurate.

A test case I've written that deals with this is https://github.com/openzipkin/zipkin/blob/master/zipkin-collector/kafka10/src/test/java/zipkin/collector/kafka10/KafkaCollectorTest.java#L135. In that test case, `receivedSpans` is a blocking queue and the Kafka consumer polling loop is buried in the KafkaCollector, but the end result should be the same.